### PR TITLE
Maps NALD charge versions to WRLS format

### PR DESCRIPTION
**Pull request by james-siteclick**
* Introduces new mapping to insert non-chargeable charge versions between gaps in NALD data
* New mapping of charge version statuses to match those in WRLS
* Removes integrity check since mapping logic has changed
* Does not import draft charge versions
* Removes redundant queries for charge agreements import

### CODE REVIEW

File: src/modules/charging-import/lib/queries/change-reasons.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+'use strict';
+
+const getNALDGapChangeReason = `
+select * from water.change_reasons where description='NALD gap'
```
Comment (Stephan-Hough) [inline src/modules/charging-import/lib/queries/change-reasons.js:4]: probably cleaner to add the semi colon at the end for the sql

File: src/modules/charging-import/lib/queries/charging.js [hunk 2]
```diff
@@ -141,43 +77,62 @@ WHERE charge_element_id IN (
...
+FROM import."NALD_CHG_VERSIONS" v
+JOIN import."NALD_ABS_LICENCES" l ON v."AABL_ID"=l."ID" AND v."FGAC_REGION_CODE"=l."FGAC_REGION_CODE"
+JOIN crm_v2.invoice_accounts ia ON ia.invoice_account_number=v."AIIA_IAS_CUST_REF" 
+JOIN import."NALD_LH_ACCS" lha ON v."AIIA_ALHA_ACC_NO"=lha."ACC_NO" AND v."FGAC_REGION_CODE"=lha."FGAC_REGION_CODE"
+JOIN crm_v2.companies c ON c.external_id=concat_ws(':', lha."FGAC_REGION_CODE", lha."ACON_APAR_ID")
+JOIN water.licences wl ON l."LIC_NO"=wl.licence_ref 
+WHERE v."FGAC_REGION_CODE"=$1 and v."AABL_ID"=$2 and v."STATUS"<>'DRAFT'
+ORDER BY 
+  to_date(v."EFF_ST_DATE", 'DD/MM/YYYY'),
+  v."VERS_NO"::integer
```
Comment (Stephan-Hough) [inline src/modules/charging-import/lib/queries/charging.js:109]: and missing semi colon here

File: src/modules/charging-import/lib/queries/charging.js [hunk 3]
```diff
@@ -141,43 +77,62 @@ WHERE charge_element_id IN (
...
+  source=EXCLUDED.source,
+  version_number=EXCLUDED.version_number,
+  invoice_account_id=EXCLUDED.invoice_account_id,
+  company_id=EXCLUDED.company_id,
+  billed_upto_date=EXCLUDED.billed_upto_date,
+  error=EXCLUDED.error,
+  scheme=EXCLUDED.scheme,
+  apportionment=EXCLUDED.apportionment,
+  change_reason_id=EXCLUDED.change_reason_id,
+  date_updated=NOW()
```
Comment (Stephan-Hough) [inline src/modules/charging-import/lib/queries/charging.js:132]: missing semi colon here

File: src/modules/charging-import/lib/queries/licences.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+l."ID",
+to_date(l."ORIG_EFF_DATE", 'DD/MM/YYYY') as start_date,
+least(
+  to_date(nullif(l."EXPIRY_DATE", 'null'), 'DD/MM/YYYY'),
+  to_date(nullif(l."REV_DATE", 'null'), 'DD/MM/YYYY'),
+  to_date(nullif(l."LAPSED_DATE", 'null'), 'DD/MM/YYYY')
+) as end_date
+FROM import."NALD_ABS_LICENCES" l 
+JOIN water.licences wl on l."LIC_NO"=wl.licence_ref
+WHERE "ORIG_EFF_DATE"<>'null'
```
Comment (Stephan-Hough) [inline src/modules/charging-import/lib/queries/licences.js:17]: missing semi colon here

File: src/modules/charging-import/mappers/charge-versions.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+ * @param {Array} source array
+ * @return {Object}
+ */
+const mapChargeVersionEndDate = (chargeVersion, index, arr) => {
+  // If there is no following row, or this row is an error, don't modify date
+  const nextRow = arr[index + 1];
+  if (!nextRow || isErrorChargeVersion(chargeVersion, nextRow)) {
+    return chargeVersion;
+  }
+  const maxEndDate = moment(nextRow.start_date, DATE_FORMAT).subtract(1, PERIOD_DAY).format(DATE_FORMAT);
```
Comment (Stephan-Hough) [inline src/modules/charging-import/mappers/charge-versions.js:59]: I haven't seen it but you might have missed handling charge versions that are invalid in NALD where the start and end dates are the same and the start date is the same as the next charge version start date. If this is the case the line above will set the end date a day before the start date.

Comment (james-siteclick) [reply src/modules/charging-import/mappers/charge-versions.js:59]: I have added an additional unit test case for this scenario and it seemed to pass OK.
I also checked the water.charge_versions table from having run locally to check if there are any with the end date before the start, and this seems OK too.

**Review by Stephan-Hough — APPROVED** [review]

**Pull request merged**